napari/_qt/_qplugins/_qnpe2.py

```diff
@@ -0,0 +1,476 @@
+    """Register submenu and actions for all npe1 plugins, clearing all first."""
+    app = get_app()
+    # Unregister all existing npe1 sample menu actions and submenus
+    if unreg := plugin_manager._unreg_sample_submenus:
+        unreg()
+    if unreg := plugin_manager._unreg_sample_actions:
+        unreg()
+
+    sample_actions: List[Action] = []
+    for plugin_name, samples in plugin_manager._sample_data.items():
+        multiprovider = len(samples) > 1
+        if multiprovider:
+            submenu_id = f'napari/file/samples/{plugin_name}'
+            submenu = [
+                (
+                    MenuId.FILE_SAMPLES,
+                    SubmenuItem(
+                        submenu=submenu_id, title=trans._(plugin_name)
+                    ),
+                ),
+            ]
+        else:
+            submenu_id = MenuId.FILE_SAMPLES
+            submenu = []
+
+        for sample_name, sample_dict in samples.items():
+
+            def _add_sample(
+                qt_viewer: QtViewer,
+                plugin: str = plugin_name,
+                sample: str = sample_name,
+            ) -> None:
+                from napari._qt.dialogs.qt_reader_dialog import (
+                    handle_gui_reading,
+                )
+
+                try:
+                    qt_viewer.viewer.open_sample(plugin, sample)
+                except MultipleReaderError as e:
+                    handle_gui_reading(
+                        [str(p) for p in e.paths],
+                        qt_viewer,
+                        stack=False,
+                    )
+
+            display_name = sample_dict['display_name'].replace('&', '&&')
+            if multiprovider:
+                title = display_name
+            else:
+                title = menu_item_template.format(plugin_name, display_name)
+
+            action: Action = Action(
+                id=f'{plugin_name}:{display_name}',
+                title=title,
+                menus=[{'id': submenu_id, 'group': MenuGroup.NAVIGATION}],
+                callback=_add_sample,
+            )
+            sample_actions.append(action)
+
+        unreg_sample_submenus = app.menus.append_menu_items(submenu)
+        plugin_manager._unreg_sample_submenus = unreg_sample_submenus
+        unreg_sample_actions = app.register_actions(sample_actions)
+        plugin_manager._unreg_sample_actions = unreg_sample_actions
+
+
+# TODO: This should be deleted once npe1 is no longer supported.
+def _toggle_or_get_widget_npe1(
+    plugin: str,
+    widget_name: str,
+    name: str,
+    hook_type: str,
+) -> None:
+    """Toggle if widget already built otherwise return widget for npe1."""
+    viewer = _provide_viewer()
+    if viewer is None:
+        raise RuntimeError(  # pragma: no cover
+            trans._(
+                'No current `Viewer` found. Note that widgets cannot be opened in headless mode.',
+                deferred=True,
+            )
+        )
```

Comment:
This show why we require a provider that raises exception. Otherwise, we raise the exception in multiple places.

Reply:
see https://github.com/napari/napari/pull/4991#discussion_r1513134397

Reply:
Stil could have a separate function that will return viewer or raise.

Reply:
So I can add a 'raise' param with default, which will hopefully allow it to work as a provider that never raises and outside where it will raise.

I do note though that I provide more detailed error message as my use case here is for widgets. I think I will allow 'raise' to take a string so I can tailor the error message.

Reply:
So I have made 2 functions as suggested. I will add test for `_provide_viewer_or_raise` as there exists test for `_provide_viewer`.

There are no current tests for `_provide_window` and `_provide_qt_viewer` - I can add them in this PR or separate PR. I know you raised concern on size of this PR. Let me know which you prefer.

Reply:
ping @Czaki

Reply:
There are no tests, but they are somehow covered. But version with exception are not covered. 
I have no worries about the size of tests

Reply:
Done